### PR TITLE
Add Caikit Standalone template

### PR DESCRIPTION
**Pull request by christinaexyou**
Closes https://issues.redhat.com/browse/RHOAIENG-6856 --> Support caikit-nlp embeddings service
Closes https://issues.redhat.com/browse/RHOAIENG-8157 --> Enable metrics for caikit standalone

## Description
Add Caikit Standalone runtime template to ODH and RHOAI

### Tested updated odh-model-controller image
1. Edited the RHOAI DSC:
```
spec:
  components:
    codeflare:
      managementState: Removed
    kserve:
      devFlags:
        manifests:
          - contextDir: config
            sourcePath: ''
            uri: 'https://github.com/christinaexyou/odh-model-controller/tarball/add-caikit-standalone'
      managementState: Managed
      serving:
        ingressGateway:
          certificate:
            secretName: knative-serving-cert
            type: SelfSigned
        managementState: Managed
        name: knative-serving
    trustyai:
      managementState: Removed
    ray:
      managementState: Removed
    kueue:
      managementState: Removed
    workbenches:
      managementState: Removed
    dashboard:
      managementState: Managed
    modelmeshserving:
      managementState: Removed
    datasciencepipelines:
      managementState: Removed
```
2. Ensured that pods within `redhat-ods-applications` were reconciled and running:
```
 oc get pods -n redhat-ods-applications        
NAME                                         READY   STATUS    RESTARTS   AGE
kserve-controller-manager-64fc4858f9-4tjxh   1/1     Running   0          14m
odh-model-controller-675f76987b-fcrj6        1/1     Running   0          14m
odh-model-controller-675f76987b-lg75n        1/1     Running   0          14m
odh-model-controller-675f76987b-r24vd        1/1     Running   0          14m
rhods-dashboard-6f64f58c89-5pgxh             2/2     Running   0          14m
rhods-dashboard-6f64f58c89-9vkjw             2/2     Running   0          14m
rhods-dashboard-6f64f58c89-d78t5             2/2     Running   0          14m
rhods-dashboard-6f64f58c89-dg5gm             2/2     Running   0          14m
rhods-dashboard-6f64f58c89-tbkh9             2/2     Running   0          14m
```
3. Verified that the `odh-model-controller` image tag was `pr-218`
```
oc get pods -n redhat-ods-applications -o jsonpath="{..image}" | tr -s '[[:space:]]' '\n' |sort |uniq -c
   2 quay.io/opendatahub/kserve-controller:latest
   6 quay.io/opendatahub/odh-model-controller:pr-218
  10 registry.redhat.io/openshift4/ose-oauth-proxy@sha256:ab112105ac37352a2a4916a39d6736f5db6ab4c29bad4467de8d613e80e9bb33
  10 registry.redhat.io/rhoai/odh-dashboard-rhel8@sha256:3a5f21a8aeefee80de9212415592273bf1e6d8542b9b1eda4ebe0c2903dbfb1f
```
4. Confirmed that `Caikit Standalone ServingRuntime` was visible in the UI
Screenshot 2024-06-04 at 4 26 07 PM

### Validated inference responses with local model outputs
* Tested via RHOAI with CPUs using the instructions [here](https://github.com/christinaexyou/caikit-embeddings/blob/main/tests/test_embeddings.py), with each of the following models: 
   * all-minilm-l12-v2
   * bge-large-en-v1.5
   * multilingual-e5-large
*  For each model, the following steps were taken:
1. Bootstrapped the model using the Caikit Embeddings module
2. Containerized the model
6. Deployed MinIO image with a reference to the containerized model
7. Deployed the model using the following SR and ISVC
```
apiVersion: serving.kserve.io/v1alpha1
kind: ServingRuntime
metadata:
  name: caikit-standalone-runtime
spec:
  multiModel: false
  supportedModelFormats:
    # Note: this currently *only* supports caikit format models
    - autoSelect: true
      name: caikit
  containers:
    - name: kserve-container
      image: quay.io/opendatahub/caikit-nlp:fast
      command: ["python", "-m", "caikit.runtime"]
      env:
        - name: RUNTIME_LOCAL_MODELS_DIR
          value: /mnt/models
        - name: TRANSFORMERS_CACHE
          value: /tmp/transformers_cache
        - name: RUNTIME_GRPC_ENABLED
          value: "false"
        - name: RUNTIME_HTTP_ENABLED
          value: "true"
      ports:
        - containerPort: 8080
          protocol: TCP
      # resources: # configure as required
      #   requests:
      #     cpu: 8
      #     memory: 16Gi
      readinessProbe:
        exec:
          command:
            - python
            - -m
            - caikit_health_probe
            - readiness
        initialDelaySeconds: 5 # might require larger values for large models
      livenessProbe:
        exec:
          command:
            - python
            - -m
            - caikit_health_probe
            - liveness
        initialDelaySeconds: 5
```
```
apiVersion: serving.kserve.io/v1alpha1
kind: ServingRuntime
metadata:
  name: caikit-standalone-runtime
spec:
  multiModel: false
  supportedModelFormats:
    # Note: this currently *only* supports caikit format models
    - autoSelect: true
      name: caikit
  containers:
    - name: kserve-container
      image: quay.io/opendatahub/caikit-nlp:fast
      command: ["python", "-m", "caikit.runtime"]
      env:
        - name: RUNTIME_LOCAL_MODELS_DIR
          value: /mnt/models
        - name: TRANSFORMERS_CACHE
          value: /tmp/transformers_cache
        - name: RUNTIME_GRPC_ENABLED
          value: "false"
        - name: RUNTIME_HTTP_ENABLED
          value: "true"
      ports:
        - containerPort: 8080
          protocol: TCP
      # resources: # configure as required
      #   requests:
      #     cpu: 8
      #     memory: 16Gi
      readinessProbe:
        exec:
          command:
            - python
            - -m
            - caikit_health_probe
            - readiness
        initialDelaySeconds: 5 # might require larger values for large models
      livenessProbe:
        exec:
          command:
            - python
            - -m
            - caikit_health_probe
            - liveness
        initialDelaySeconds: 5
```
8. Validated the 6 different embedding inference responses with a local model, using the test script [here](https://github.com/christinaexyou/caikit-embeddings/blob/main/tests/test_embeddings.py). Here is an example inference request to the `/api/v1/task/rerank/` endpoint for the model `all-minilm-l12-v2`:
```
>> inference_response = requests.post(
            isvc_url + "/api/v1/task/rerank",
            json={
             "model_id": model_id,
             "inputs": {
                 "documents": documents,
                 "query": text
             },
            # "parameters": {
            #     "top_n":
            #     }
            },
            verify=False
        ).json()
>> print(inference_response['result'])
{'query': 'test first sentence',
 'scores': [{'document': {'text': 'first sentence', 'title': 'first title'},
   'index': 0,
   'score': 0.6441316604614258,
   'text': 'first sentence'},
  {'document': {'text': 'another sentence', 'more': 'more attributes here'},
   'index': 1,
   'score': 0.4829946458339691,
   'text': 'another sentence'},
  {'document': {'text': 'a doc with a nested metadata',
    'meta': {'foo': 'bar', 'i': 999, 'f': 12.34}},
   'index': 2,
   'score': 0.0344165600836277,
   'text': 'a doc with a nested metadata'}]}
```

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: config/runtimes/caikit-standalone-template.yaml [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    metadata:
+      name: caikit-standalone-runtime
+      annotations:
+        openshift.io/display-name: Caikit Standalone ServingRuntime for KServe
+        opendatahub.io/recommended-accelerators: '["nvidia.com/gpu"]'
+      labels:
+        opendatahub.io/dashboard: 'true'
+    spec:
+      annotations:
+        prometheus.io/port: '3000'
```
Comment (VedantMahabaleshwarkar) [inline config/runtimes/caikit-standalone-template.yaml:29]: IIRC this was 8086 for caikit. But I'm not sure if that changed with this build. Can we verify what port caikit metrics are emitted on and ensure this annotation has the correct value?

File: config/base/params.env [hunk 2]
```diff
@@ -1,5 +1,6 @@
-odh-model-controller=quay.io/opendatahub/odh-model-controller:fast
+odh-model-controller=quay.io/opendatahub/odh-model-controller:pr-218
```
Comment (israel-hdez) [inline config/base/params.env:1]: Don't forget to revert this line.

File: config/runtimes/caikit-standalone-template.yaml (hunk 1 above)
Comment (dtrifiro) [reply config/runtimes/caikit-standalone-template.yaml:29]: That's correct, this needs to be 8086

File: config/runtimes/caikit-standalone-template.yaml [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+          image: $(caikit-standalone-image)
+          command:
+            - python
+            - '-m'
+            - caikit.runtime
+          env:
+            - name: RUNTIME_LOCAL_MODELS_DIR
+              value: /mnt/models
+            - name: TRANSFORMERS_CACHE
+              value: /tmp/transformers_cache
```
Comment (dtrifiro) [inline config/runtimes/caikit-standalone-template.yaml:46]: This has been deprecated and can be safely replaced with this:

```suggestion
            - name: HF_HOME
              value: /tmp/hf_home
```

**Comment by christinaexyou** [comment]
@VedantMahabaleshwarkar, @israel-hdez, @dtrifiro thanks for your review ! the following changes were made: 
- set the prometheus port to 8086
- replaced `TRANSFORMERS_CACHE` key and value with `HF_HOME`

File: controllers/constants/caikit-metrics.json [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    "metrics": {
+        "supported": "true",
+        "config": [
+            {
+                "title": "Number of requests",
+                "type": "REQUEST_COUNT",
+                "queries": [
+                    {
+                        "title": "Number of successful incoming requests",
+                        "query": "sum(increase(predict_rpc_count_total{namespace='${NAMESPACE}',code='OK',model_id='${MODEL_ID}',grpc_request='${TASK}'}[${RATE_INTERVAL}]))"
```
Comment (VedantMahabaleshwarkar) [inline controllers/constants/caikit-metrics.json:11]: what does the `${TASK}` variable need to be replaced with while making the final query? @christinaexyou

Comment (christinaexyou) [reply controllers/constants/caikit-metrics.json:11]: The specific embeddings task, e.g. EmbeddingTaskRequest, SentenceSimilarityTaskRequest

**Review by dtrifiro — APPROVED** [review]
Looks good to me, just needs DCO signoff:

```bash
git rebase upstream/main -x 'git commit --amend --no-edit --signoff'
```

File: controllers/constants/caikit-metrics.json [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    "metrics": {
+        "supported": "true",
+        "config": [
+            {
+                "title": "Number of requests",
+                "type": "REQUEST_COUNT",
+                "queries": [
+                    {
+                        "title": "Number of successful incoming requests",
+                        "query": "sum(increase(predict_rpc_count_total{namespace='${NAMESPACE}',code='OK',model_id='${MODEL_ID}'}[${RATE_INTERVAL}]))"
```
Comment (VedantMahabaleshwarkar) [inline controllers/constants/caikit-metrics.json:11]: I missed this earlier, but can we change `${MODEL_ID}` to `${MODEL_NAME}`? The replacement logic is written for `${MODEL_NAME}`

**Comment by dtrifiro** [comment]
@Jooho this PR is in the [release-0.12.0](https://github.com/opendatahub-io/odh-model-controller/tree/release-0.12.0) branch but missing in [`v0.12.2`](https://github.com/opendatahub-io/odh-model-controller/blob/v0.12.0.2/config/base/params.env) (side note: github shows this warning: `This commit does not belong to any branch on this repository, and may belong to a fork outside of the repository.` for this tag and `v0.12.0` and `v0.12.1`).

Can we release `v0.12.0.3` to include this? 

cc @christinaexyou

**Pull request merged**